File: packages/mobile/android/app/src/main/java/id/homebase/lib/core/file/CryptoUtil.java
package id.homebase.lib.core.file;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.charset.StandardCharsets;
import java.security.GeneralSecurityException;
import java.util.Base64;
import java.util.concurrent.CompletableFuture;

import javax.crypto.Cipher;
import javax.crypto.SecretKey;
import javax.crypto.spec.IvParameterSpec;
import javax.crypto.spec.SecretKeySpec;

import id.homebase.lib.core.DotYouClient;
import okhttp3.MediaType;
import okhttp3.RequestBody;
import okio.BufferedSink;
import okio.Okio;

public class CryptoUtil {

    public static SecretKey importKey(byte[] keyBytes) {
        return new SecretKeySpec(keyBytes, "AES");
    }

    public static byte[] innerEncrypt(byte[] iv, SecretKey key, byte[] data) throws Exception {
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        cipher.init(Cipher.ENCRYPT_MODE, key, new IvParameterSpec(iv));
        return cipher.doFinal(data);
    }

    public static byte[] cbcEncrypt(byte[] data, byte[] iv, byte[] key) throws Exception {
        SecretKey secretKey = importKey(key);
        return innerEncrypt(iv, secretKey, data);
    }

    public static ByteArrayOutputStream cbcEncryptStream(File inputFile, byte[] iv, byte[] key) throws Exception {
        SecretKeySpec secretKey = new SecretKeySpec(key, "AES");
        Cipher cipher = Cipher.getInstance("AES/CBC/PKCS5Padding");
        IvParameterSpec ivSpec = new IvParameterSpec(iv);
        cipher.init(Cipher.ENCRYPT_MODE, secretKey, ivSpec);

        try (FileInputStream fis = new FileInputStream(inputFile);
             ByteArrayOutputStream baos = new ByteArrayOutputStream()) {

            byte[] buffer = new byte[16]; // Block size for AES

            while (true) {
                int bytesRead = fis.read(buffer);
                if (bytesRead == -1) {
                    break; // End of file
                }

                byte[] encryptedBytes = cipher.update(buffer, 0, bytesRead);
                baos.write(encryptedBytes);
            }

            // Write the last block of encrypted data (with padding)
            byte[] finalEncryptedBytes = cipher.doFinal();
            baos.write(finalEncryptedBytes);

            return baos;

        } catch (IOException e) {
            throw new RuntimeException("Error reading file", e);
        }
    }


    // Method to convert a string to a byte array
    public static byte[] stringToByteArray(String input) {
        return input.getBytes(StandardCharsets.UTF_8);
    }

    // Method to convert a byte array to a base64 string
    public static String byteArrayToBase64(byte[] input) {
        return Base64.getEncoder().encodeToString(input);
    }

    // Encrypt metadata method
    public static UploadFileMetadata<String> encryptMetaData(UploadFileMetadata<String> metadata, KeyHeader keyHeader) throws Exception {
        if (keyHeader != null && metadata.getAppData().content() != null) {
            String encryptedContent = encryptWithKeyheader(metadata.getAppData().content(), keyHeader);
            metadata.getAppData().content(encryptedContent);
        }
        return metadata;
    }

    // Encrypt with key header method
    public static String encryptWithKeyheader(String content, KeyHeader keyHeader) throws Exception {
        byte[] contentBytes = stringToByteArray(content);
        byte[] encryptedBytes = encryptWithKeyheader(contentBytes, keyHeader);
        return byteArrayToBase64(encryptedBytes);
    }

    public static byte[] encryptWithKeyheader(byte[] contentBytes, KeyHeader keyHeader) throws Exception {
        return cbcEncrypt(contentBytes, keyHeader.iv(), keyHeader.aesKey());
    }

    public static ByteArrayOutputStream encryptWithKeyheader(File contentBytes, KeyHeader keyHeader) throws Exception {
        return cbcEncryptStream(contentBytes, keyHeader.iv(), keyHeader.aesKey());

    }


    private static byte[] concatenateByteArrays(byte[] array1, byte[] array2) {
        byte[] combined = new byte[array1.length + array2.length];
        System.arraycopy(array1, 0, combined, 0, array1.length);
        System.arraycopy(array2, 0, combined, array1.length, array2.length);
        return combined;
    }

    public static EncryptedKeyHeader encryptKeyHeader(
            DotYouClient dotYouClient,
            KeyHeader keyHeader,
            byte[] transferIv) throws Exception {

        byte[] ss = dotYouClient.getSharedSecret();
        if (ss == null) {
            throw new IllegalStateException("Attempting to encrypt but missing the shared secret");
        }

        byte[] combined = concatenateByteArrays(keyHeader.iv(), keyHeader.aesKey());
        byte[] encryptedAesKey = cbcEncrypt(combined, transferIv, ss);
        return new EncryptedKeyHeader(1, 11, transferIv, encryptedAesKey);
    }
}

File: packages/mobile/ios/core/CryptoUtil.swift
//
//  CryptoUtil.swift
//  HomebasePhotos
//
//  Created by Stef Coenen on 21/06/2024.
//

import Foundation
import CryptoKit
import CommonCrypto

class CryptoUtil {
  static func encryptKeyHeader(dotYouClient: DotYouClient, keyHeader: KeyHeader, transferIv: Data) throws -> EncryptedKeyHeader {
      guard let sharedSecret = dotYouClient.sharedSecret else {
          throw NSError(domain: "MissingSharedSecret", code: -1, userInfo: [NSLocalizedDescriptionKey: "Attempting to encrypt but missing the shared secret"])
      }

      // Combine keyHeader.iv and keyHeader.aesKey
      var combined = Data()
      combined.append(keyHeader.iv)
      combined.append(keyHeader.aesKey)

      // Perform the encryption using cbcEncrypt
      let cipher = try cbcEncrypt(data: combined, iv: transferIv, key: sharedSecret)

      return EncryptedKeyHeader(
          iv: transferIv,
          encryptedAesKey: cipher,
          encryptionVersion: 1,
          type: 11
      )
  }


  static func encryptMetaData(metadata: UploadFileMetadata<String>, keyHeader: KeyHeader ) throws -> UploadFileMetadata<String> {
    if(metadata.appData.content == nil){
      return metadata;
    }
    var encryptedMetadata = metadata

    print("encryptMetaData", encryptedMetadata.appData.content)
    
    let encryptedContent = try innerEncrypt(iv:keyHeader.iv, key: keyHeader.aesKey, data: metadata.appData.content.data(using: .utf8)!)
    encryptedMetadata.appData.content = encryptedContent.base64EncodedString()

    return encryptedMetadata;
  }

  static func encryptWithKeyHeader(file: String, keyHeader: KeyHeader) throws -> (stream: InputStream, count: UInt64) {
    guard let inputStream = InputStream(fileAtPath: file) else {
      throw NSError(domain: "Invalid file as stream", code: -1, userInfo: nil)
    }

    return try encryptWithKeyHeader(stream: inputStream, keyHeader: keyHeader)
  }

  static func encryptWithKeyHeader(stream: InputStream, keyHeader: KeyHeader) throws -> (stream: InputStream, count: UInt64) {
    let keyData = keyHeader.aesKey
    var iv = keyHeader.iv

    let outputStream = OutputStream(toMemory: ())

    stream.open()
    outputStream.open()

    defer {
      stream.close()
      outputStream.close()
    }

    let blockSize = kCCBlockSizeAES128
    var buffer = [UInt8](repeating: 0, count: blockSize)

    var lastPadding = iv

    while stream.hasBytesAvailable {
      let bytesRead = stream.read(&buffer, maxLength: blockSize)
      if bytesRead == 0 { break }

      let inputData = Data(bytes: buffer, count: bytesRead)
      let bufferSize = inputData.count + blockSize
      var encryptedBuffer = [UInt8](repeating: 0, count: bufferSize)

      var numBytesEncrypted: size_t = 0
      let result = encryptedBuffer.withUnsafeMutableBytes { encryptedBytes in
        CCCrypt(CCOperation(kCCEncrypt),
                CCAlgorithm(kCCAlgorithmAES),
                CCOptions(kCCOptionPKCS7Padding),
                keyData.withUnsafeBytes { $0.baseAddress },
                keyData.count,
                iv.withUnsafeBytes { $0.baseAddress },
                inputData.withUnsafeBytes { $0.baseAddress },
                inputData.count,
                encryptedBytes.baseAddress,
                bufferSize,
                &numBytesEncrypted)
      }

      guard result == kCCSuccess else {
        throw NSError(domain: "Encryption failed", code: Int(result), userInfo: nil)
      }


      // Get the last 16 bytes (padding) of the encrypted data
      lastPadding = Data(encryptedBuffer[(numBytesEncrypted - blockSize)..<numBytesEncrypted])

      // Remove the padding from the encrypted data
      numBytesEncrypted -= blockSize

      if(numBytesEncrypted == 0) {
        // Appaerently this is expected at the end... :shrug:
        break;
      } else {
        // Copy the last block of the encrypted data into the iv (without padding)
        iv = Data(encryptedBuffer[(numBytesEncrypted - blockSize)..<numBytesEncrypted])
      }

      // Write the encrypted data to the output stream
      let encryptedData = Data(bytes: encryptedBuffer, count: numBytesEncrypted)
      encryptedData.withUnsafeBytes { outputStream.write($0.bindMemory(to: UInt8.self).baseAddress!, maxLength: numBytesEncrypted) }
    }

    // Write the last padding
    lastPadding.withUnsafeBytes {
      outputStream.write($0.bindMemory(to: UInt8.self).baseAddress!, maxLength: blockSize)
    }

    let encryptedData = outputStream.property(forKey: .dataWrittenToMemoryStreamKey) as! Data
    return (stream: InputStream(data: encryptedData), count: UInt64(encryptedData.count))
  }

  static func innerEncrypt(iv: Data, key: Data, data: Data) throws -> Data {
    let keyLength = kCCKeySizeAES128
    let ivLength = kCCBlockSizeAES128
    let options = CCOptions(kCCOptionPKCS7Padding)

    var encryptedData = Data(count: data.count + ivLength)
    var numBytesEncrypted = 0

    // Create a local copy of the buffer to avoid overlapping access issues
    var tempEncryptedData = encryptedData

    let cryptStatus = tempEncryptedData.withUnsafeMutableBytes { encryptedBytes in
        data.withUnsafeBytes { dataBytes in
            iv.withUnsafeBytes { ivBytes in
                key.withUnsafeBytes { keyBytes in
                    CCCrypt(CCOperation(kCCEncrypt),
                            CCAlgorithm(kCCAlgorithmAES),
                            options,
                            keyBytes.baseAddress, keyLength,
                            ivBytes.baseAddress,
                            dataBytes.baseAddress, data.count,
                            encryptedBytes.baseAddress, encryptedData.count,
                            &numBytesEncrypted)
                }
            }
        }
    }

    guard cryptStatus == kCCSuccess else {
        throw NSError(domain: "Error in encryption", code: Int(cryptStatus), userInfo: nil)
    }

    // Update the original encryptedData only after the crypt operation
    encryptedData = tempEncryptedData
    encryptedData.removeSubrange(numBytesEncrypted..<encryptedData.count)
    return encryptedData
}

  static func cbcEncrypt(data: Data, iv: Data, key: Data) throws -> Data {
      return try innerEncrypt(iv: iv, key: key, data: data)
  }

  static func getUpdatedKeyHeader(keyHeader: KeyHeader, manifest: UploadManifest, payloadKey: String) -> KeyHeader {
    let manifest = manifest;
    let payloadDescriptors = manifest.payloadDescriptors;

    for descriptor in payloadDescriptors {
        if descriptor.payloadKey == payloadKey {
          if(descriptor.iv != nil) {
            return KeyHeader(iv: descriptor.iv!, aesKey: keyHeader.aesKey)
          }
        }
    }

    return keyHeader
  }

  static func base64Encode(stream: InputStream) -> String? {
      stream.open()
      defer {
          stream.close()
      }

      let bufferSize = 1024
      var buffer = [UInt8](repeating: 0, count: bufferSize)
      var data = Data()

      while stream.hasBytesAvailable {
          let read = stream.read(&buffer, maxLength: bufferSize)
          if read < 0 {
              print("Error reading stream: \(stream.streamError?.localizedDescription ?? "Unknown error")")
              return nil
          } else if read == 0 {
              break
          }
          data.append(buffer, count: read)
      }

      return data.base64EncodedString()
  }

}
